docs/src/writing_good_rules.md

```diff
@@ -103,16 +103,197 @@ In fact, any number of scalar arguments is supported, as is returning a tuple of
 See docstrings for the comprehensive usage instructions.
 ## Write tests
 
-In [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
-there are fairly decent tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
-Take a look at existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests and you should see how to do stuff.
+[ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
+provides tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+Take a look at the documentation or the existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests to see how to write the tests.
+It's easy!
 
 !!! warning
-    Use finite differencing to test derivatives.
     Don't use analytical derivations for derivatives in the tests.
     Those are what you use to define the rules, and so can not be confidently used in the test.
     If you misread/misunderstood them, then your tests/implementation will have the same mistake.
+    Use finite differencing methods instead, as they are based on the primal computation.
 
 ## CAS systems are your friends.
 
 It is very easy to check gradients or derivatives with a computer algebra system (CAS) like [WolframAlpha](https://www.wolframalpha.com/input/?i=gradient+atan2%28x%2Cy%29).
+
+## Which functions need rules?
+
+In principle, a perfect AD system only needs rules for basic operations and can infer the
+rules for more complicated functions automatically.
+In practice, performance needs to be considered as well.
+
+Some functions use `ccall` internally for performance reasons, for example [`^`]
+(https://github.com/JuliaLang/julia/blob/380abc808ad8fbd72a7a57f6ac0f811b7903046d/base/math.jl#L922).
+These functions can not be differentiated through by AD systems, and need custom rules.
```

Comment:
Is this actually true? IRC Enzyme works at llvm level so might be able to do these?

Reply:
@oxinabox your wisdom is required

Reply:
2 points:
1.) It isn't normally for performance reasons. For a lot of mathematical functions its just that it hasn't been ported to julia yet (e.g. `FastMath`), because porting it is a lot of work for litttle gain. For a lot of other things, like allocating and resizing arrays, it is expressing a fundermental operation that doesn't exist in julia. Also all built-ins

2.) Enzyme *can* do these things, but it is a special case. We are talking at a general enough level. In general AD systems can not cross the language binaries, nor AD things that are too primative (fundermentally one at very least has to hit things that has assembly instructions, since there is nothing to decompose). We should not worry about being so strictly true that this becomes hard to read